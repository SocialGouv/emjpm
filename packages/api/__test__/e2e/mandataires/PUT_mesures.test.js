//

const request = require("supertest");

// Fake env
process.env.SMTP_FROM = "ne-pas-repondre@emjpm.gouv.fr";
process.env.APP_URL = "https://emjpm.gouv.fr";

const server = require("@emjpm/api/app");
const knex = require("@emjpm/api/db/knex");

const { getTokenByUserType } = require("../utils");

beforeEach(async () => {
  await knex.migrate.latest();
  await knex.seed.run();
});

afterAll(async () => {
  await knex.destroy();
});

const sampleMesure = {
  code_postal: "28000",
  ville: "Chartres",
  etablissement: "ROOKIE",
  created_at: "2010-10-05",
  annee: "2010-10-05",
  type: "preposes",
  date_ouverture: "2010-10-05",
  residence: "oui",
  civilite: "madame",
  status: "Mesure en cours"
};

const getMesuresCount = async id =>
  (await knex("mandataires")
    .select("mesures_en_cours")
    .where({
      id
    })
    .first()).mesures_en_cours;

test("mandataire should PUT his mesure", async () => {
  const token = await getTokenByUserType("mandataire");

  const newMesureId = await knex("mesures")
    .insert({
      ...sampleMesure,
      mandataire_id: 1,
      ti_id: 1
    })
    .returning("id")
    .then(res => res[0]);

  const response = await request(server)
    .put(`/api/v1/mandataires/1/mesures/${newMesureId}`)
    .set("Authorization", "Bearer " + token)
    .send({
      etablissement: "DRUID"
    });

  const newMesure = await knex("mesures")
    .where({ id: newMesureId })
    .first();

  expect(newMesure.etablissement).toBe("DRUID");

  expect(response.body).toMatchSnapshot();
  expect(response.status).toBe(200);
});

test("mandataire should NOT PUT other mandataire mesure", async () => {
  const token = await getTokenByUserType("mandataire");

  const newMesureId = await knex("mesures")
    .insert({
      ...sampleMesure,
      mandataire_id: 2,
      ti_id: 1
    })
    .returning("id")
    .then(res => res[0]);

  const response = await request(server)
    .put(`/api/v1/mandataires/2/mesures/${newMesureId}`)
    .set("Authorization", "Bearer " + token)
    .send({
      etablissement: "DRUID"
    });

  const newMesure = await knex("mesures")
    .where({ id: newMesureId })
    .first();

  expect(newMesure.etablissement).toBe("ROOKIE");

  expect(response.body).toMatchSnapshot();
  expect(response.status).toBe(200);
});

test("TI should PUT his mandataire mesure", async () => {
  const token = await getTokenByUserType("ti");

  const newMesureId = await knex("mesures")
    .insert({
      ...sampleMesure,
      mandataire_id: 1,
      ti_id: 2
    })
    .returning("id")
    .then(res => res[0]);

  const response = await request(server)
    .put(`/api/v1/mandataires/1/mesures/${newMesureId}`)
    .set("Authorization", "Bearer " + token)
    .send({
      id: newMesureId,
      etablissement: "DRUID"
    });

  const newMesure = await knex("mesures")
    .where({ id: newMesureId })
    .first();

  expect(newMesure.etablissement).toBe("DRUID");

  expect(response.body).toMatchSnapshot();
  expect(response.status).toBe(200);
});

test("TI should NOT PUT other TI mesure", async () => {
  const token = await getTokenByUserType("ti");
  const newMesureId = await knex("mesures")
    .insert({
      ...sampleMesure,
      mandataire_id: 1,
      ti_id: 1
    })
    .returning("id")
    .then(res => res[0]);

  const response = await request(server)
    .put(`/api/v1/mandataires/1/mesures/${newMesureId}`)
    .set("Authorization", "Bearer " + token)
    .send({
      id: newMesureId,
      etablissement: "DRUID"
    });

  const newMesure = await knex("mesures")
    .where({ id: newMesureId })
    .first();

  expect(newMesure.etablissement).toBe("ROOKIE");

  expect(response.body).toMatchSnapshot();
  expect(response.status).toBe(200);
});
