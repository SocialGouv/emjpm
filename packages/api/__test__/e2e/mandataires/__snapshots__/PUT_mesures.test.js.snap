// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TI should NOT PUT other TI mesure 1`] = `
Array [
  Object {
    "annee": "2010-10-05",
    "cabinet": null,
    "civilite": "F",
    "code_postal": "93000",
    "created_at": "2010-10-05T00:00:00.000Z",
    "date_ouverture": "2010-10-04T22:00:00.000Z",
    "etablissement": "UDAHF",
    "etablissement_id": null,
    "extinction": null,
    "id": 3,
    "manda": "UDAHF",
    "mandataire_id": 1,
    "numero_dossier": null,
    "residence": null,
    "status": "Mesure en attente",
    "ti_id": null,
    "type": "preposes",
    "ville": null,
  },
]
`;

exports[`TI should PUT his mandataire mesure 1`] = `
Array [
  Object {
    "annee": "2010-10-05",
    "cabinet": null,
    "civilite": "F",
    "code_postal": "93000",
    "created_at": "2010-10-05T00:00:00.000Z",
    "date_ouverture": "2010-10-04T22:00:00.000Z",
    "etablissement": "UDAHF",
    "etablissement_id": null,
    "extinction": null,
    "id": 3,
    "manda": "UDAHF",
    "mandataire_id": 1,
    "numero_dossier": null,
    "residence": null,
    "status": "Mesure en attente",
    "ti_id": null,
    "type": "preposes",
    "ville": null,
  },
]
`;

exports[`mandataire should NOT PUT other mandataire mesure 1`] = `
Array [
  Object {
    "annee": "2010-10-05",
    "cabinet": null,
    "civilite": "F",
    "code_postal": "62000",
    "created_at": "2010-10-05T00:00:00.000Z",
    "date_ouverture": "2010-10-04T22:00:00.000Z",
    "etablissement": "UDAHF",
    "etablissement_id": null,
    "extinction": null,
    "id": 1,
    "mandataire_id": 1,
    "numero_dossier": null,
    "residence": "oui",
    "status": "Mesure en cours",
    "ti_id": null,
    "type": "preposes",
    "ville": "Arras",
  },
  Object {
    "annee": "2010-10-05",
    "cabinet": null,
    "civilite": "H",
    "code_postal": "93000",
    "created_at": "2010-10-05T00:00:00.000Z",
    "date_ouverture": "2010-10-04T22:00:00.000Z",
    "etablissement": "CAROUF",
    "etablissement_id": null,
    "extinction": null,
    "id": 4,
    "mandataire_id": 1,
    "numero_dossier": null,
    "residence": "oui",
    "status": "Mesure en cours",
    "ti_id": null,
    "type": "preposes",
    "ville": "St Denis",
  },
]
`;

exports[`mandataire should PUT his mesure 1`] = `
Array [
  Object {
    "annee": "2010-10-05",
    "cabinet": null,
    "civilite": "F",
    "code_postal": "62000",
    "created_at": "2010-10-05T00:00:00.000Z",
    "date_ouverture": "2010-10-04T22:00:00.000Z",
    "etablissement": "UDAHF",
    "etablissement_id": null,
    "extinction": null,
    "id": 1,
    "mandataire_id": 1,
    "numero_dossier": null,
    "residence": "oui",
    "status": "Mesure en cours",
    "ti_id": null,
    "type": "preposes",
    "ville": "Arras",
  },
  Object {
    "annee": "2010-10-05",
    "cabinet": null,
    "civilite": "H",
    "code_postal": "93000",
    "created_at": "2010-10-05T00:00:00.000Z",
    "date_ouverture": "2010-10-04T22:00:00.000Z",
    "etablissement": "CAROUF",
    "etablissement_id": null,
    "extinction": null,
    "id": 4,
    "mandataire_id": 1,
    "numero_dossier": null,
    "residence": "oui",
    "status": "Mesure en cours",
    "ti_id": null,
    "type": "preposes",
    "ville": "St Denis",
  },
  Object {
    "annee": "2010-10-05",
    "cabinet": null,
    "civilite": "madame",
    "code_postal": "28000",
    "created_at": "2010-10-05T00:00:00.000Z",
    "date_ouverture": "2010-10-04T22:00:00.000Z",
    "etablissement": "DRUID",
    "etablissement_id": null,
    "extinction": null,
    "id": 6,
    "mandataire_id": 1,
    "numero_dossier": null,
    "residence": "oui",
    "status": "Mesure en cours",
    "ti_id": 1,
    "type": "preposes",
    "ville": "Chartres",
  },
]
`;
